### PR TITLE
DuckDuckGo Shortcuts Cheat Sheet

### PR DESCRIPTION
**Pull request by Qeole**
# DuckDuckGo Shortcuts Cheat Sheet

**What does your Instant Answer do?**
It provides a cheat sheet about keyboard shortcuts available on DuckDuckGo SERP.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Faster to access. Answers appear on the very page where it's needed.

**What is the data source for your Instant Answer? (Provide a link if possible)**
https://duck.co/help/features/keyboard-shortcuts

**Why did you choose this data source?**
Somehow, that looked sensible to me…

**Are there any other alternative (better) data sources?**
None that I know of.

**What are some example queries that trigger this Instant Answer?**
"shortcuts", "DuckDuckGo shortcuts", "shortcuts ddg", "keyboard cheat sheet shortcuts DDG"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
That would typically be for power users of DuckDuckGo

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
This one: https://duck.co/ideas/idea/4870/display-duckduckgo-keyboard-shortcuts

**Which existing Instant Answers will this one supercede/overlap with?**
None that I know of.

**Are you having any problems? Do you need our help with anything?**
Seems to work well. Feedback welcome!

**Where did you hear about DuckDuckHack? (For first time contributors)**
Sorry, can't remember. Probably an article on the web.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
Cheat sheet.

ddg1

## Checklist

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Linux (Kubuntu)
        [X] Firefox

### CODE REVIEW

File: lib/DDG/Goodie/DdgShortcuts.pm [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+name "DDG_Shortcuts";
+description "DuckDuckGo Shortcuts Cheat Sheet";
+primary_example_queries "DuckDuckGo shortcuts", "shortcuts";
+secondary_example_queries "keyboard shortcuts", "shortcuts cheatsheet";
+category "cheat_sheets";
+topics "computing", "geek", "trivia";
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/DDGShortcuts.pm";
+attribution github => "Qeole",
+            twitter => "qeole";
```
Comment (MrChrisW) [inline lib/DDG/Goodie/DdgShortcuts.pm]: While not required can you format the attribution like this: https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/GithubJobs.pm#L13-L14

File: lib/DDG/Goodie/DdgShortcuts.pm [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+triggers any => "shortcuts";
+triggers startend => "shortcuts";
+
+handle remainder => sub {
+
+    # Combination of "keyboard", "cheat sheet", "duckduckgo" (or aliases), or nothing
+    return unless $_ =~ /^\s*(((duck\s*duck\s*go|ddg)( ?'?s)?|cheat\s*sheet|keyboard)\s*)*$/i;
+
```
Comment (MrChrisW) [inline lib/DDG/Goodie/DdgShortcuts.pm]: Probably best to remove this regexp guard and simply update your `triggers` above to something like this 

https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/CrontabCheatSheet.pm#L19-L36 

Using triggers only will result in a faster operation.
Example: `ddg cheatsheet` `duckduckgo cheatsheet`

File: lib/DDG/Goodie/DdgShortcuts.pm [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    return $HTML;
+}
+
+my $TEXT;
+
+sub text_cheat_sheet {
+    $TEXT //= share("ddg_shortcuts.txt")
+        ->slurp(iomode => '<:encoding(UTF-8)');
+    return $TEXT;
+}
```
Comment (MrChrisW) [inline lib/DDG/Goodie/DdgShortcuts.pm]: @Qeole I understand you've just replicated other cheatsheets, however recently we had a discussion about changing the way we're slurping the html and txt file.  Can you update this to adhere to that change https://github.com/jim-brighter/zeroclickinfo-goodies/commit/08b7dd21a34b9c4374f046d88e01c14586ff0421

You should end up with something like this:

````
my $TEXT = scalar share('ddg_shortcuts.txt')->slurp,
my $HTML = scalar share('ddg_shortcuts.html')->slurp;
handle remainder => sub {
return
    heading => 'DuckDuckGo Shortcuts Cheat Sheet',
    answer => $TEXT,
    html => $HTML,
};
````

**Comment by MrChrisW** [comment]
Hi @Qeole thanks for your contribution! I've added some feedback, feel free to let me know if you have any questions. :cake:

File: lib/DDG/Goodie/DdgShortcuts.pm (hunk 2 above)
Comment (mintsoft) [inline lib/DDG/Goodie/DdgShortcuts.pm]: Definitely; those triggers seem way too broad to me

File: lib/DDG/Goodie/DdgShortcuts.pm [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+primary_example_queries "DuckDuckGo shortcuts", "shortcuts";
+secondary_example_queries "keyboard shortcuts", "shortcuts cheatsheet";
+category "cheat_sheets";
+topics "computing", "geek", "trivia";
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/DDGShortcuts.pm";
+attribution github => "Qeole",
+            twitter => "qeole";
+
+triggers any => "shortcuts";
+triggers startend => "shortcuts";
```
Comment (mintsoft) [inline lib/DDG/Goodie/DdgShortcuts.pm]: Also "any" covers "startend" so you don't need both

File: share/goodie/ddg_shortcuts/ddg_shortcuts.css [hunk 5]
```diff
@@ -0,0 +1,45 @@
+@media (max-width: 400px) {
+    .zci--answer .ddg_shortcuts-column {
+        width: 100%;
+    }
+}
+
+.zci--answer .zero_click_wrapper .zero_click {
+    padding-left: 0px !important;
```
Comment (mintsoft) [inline share/goodie/ddg_shortcuts/ddg_shortcuts.css]: Is this `!important` copied from the other cheatsheet CSS? I don't recall seeing that in any other goodies

File: share/goodie/ddg_shortcuts/ddg_shortcuts.css [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    margin-left:  0px !important;
+    margin-right: 0px !important;
+}
+
+.zci--answer .ddg_shortcuts-container {
+    max-height:  45ex;
+    /overflow-y: auto;
```
Comment (mintsoft) [inline share/goodie/ddg_shortcuts/ddg_shortcuts.css]: Is this an IE8 hack? `/overflow-y` If so can you comment above

File: share/goodie/ddg_shortcuts/ddg_shortcuts.css (hunk 5 above)
Comment (MrChrisW) [inline share/goodie/ddg_shortcuts/ddg_shortcuts.css]: @mintsoft Included in the [orginal VimCheatSheet](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/share/goodie/vim_cheat_sheet/vim_cheat_sheet.css#L1-L5) css.

Comment (Qeole) [inline share/goodie/ddg_shortcuts/ddg_shortcuts.css]: @minsoft I copied that from [tmux cheat sheet](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/share/goodie/tmux_cheat_sheet/tmux_cheat_sheet.css) to be precise.

File: lib/DDG/Goodie/DdgShortcuts.pm (hunk 1 above)
Comment (Qeole) [inline lib/DDG/Goodie/DdgShortcuts.pm]: Thanks, I did that in new commit.

File: lib/DDG/Goodie/DdgShortcuts.pm (hunk 2 above)
Comment (Qeole) [inline lib/DDG/Goodie/DdgShortcuts.pm]: I wrote again my triggers and removed the guard. I made again a lot of triggers, maybe there are too many; please have a look at it. The comprehensive list of (90) generated expressions is in the test file.

File: lib/DDG/Goodie/DdgShortcuts.pm (hunk 3 above)
Comment (Qeole) [inline lib/DDG/Goodie/DdgShortcuts.pm]: Right, I had not seen this indeed. It's corrected in new commit.

File: share/goodie/ddg_shortcuts/ddg_shortcuts.css (hunk 6 above)
Comment (Qeole) [inline share/goodie/ddg_shortcuts/ddg_shortcuts.css]: That's a mistake. No idea how it occured, but I definitely did not intend to add the `/`.

File: lib/DDG/Goodie/DdgShortcuts.pm (hunk 4 above)
Comment (Qeole) [inline lib/DDG/Goodie/DdgShortcuts.pm]: This is what I would have expected, but my trigger would not work with only `any` if there were no other words around it.
Maybe you could check that you have the behavior you expect on your side, and that I'm the only one to have an issue here?

**Comment by Qeole** [comment]
I've made two new commits, one with minor changes to correct style and `/` mistake in CSS.

The second one is about re-writing the triggers: I now try to catch many expressions, but without regex guard. The full list of generated triggers is available in test file. Maybe it's overkill, tell me if I should really simplify it.

And thanks for feedback!

**Comment by MrChrisW** [comment]
@Qeole that's definitely an extensive list of triggers :+1: .They all seem reasonable to me. 

I think you can remove the [debug](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/956/files#diff-8c9fd74002d653e473cf561f2a5f8ad4R39map) code.

Happy to help! Thanks for contributing.

**Comment by Qeole** [comment]
@MrChrisW Done! I just removed commented debug code.

**Comment by mintsoft** [comment]
This lgtm @jagtalon @moollaza @chrismorast any additional feedback?

**Comment by moollaza** [comment]
Everything LGTM  :+1: 

/cc @abeyang

**Comment by jagtalon** [comment]
The misaligned columns are throwing me off a bit

screen shot 2015-01-30 at 4 16 22 am

Although it might look worse if adjusted? Not sure. Also, I think it's better to use arrows instead of `Left` or `Right` because it's not clear if that's the arrow key or the mouse. Something like Asana's is pretty clear:

screen shot 2015-01-30 at 4 20 35 am

**Comment by jagtalon** [comment]
Enter could just be `↵` too to save space. What do you think @moollaza @abeyang?

**Comment by jagtalon** [comment]
Up in https://ddh1.duckduckgo.com/?q=duckduckgo+shortcuts&ia=answer

**Comment by Qeole** [comment]
Left column looked better misaligned than aligned (in my opinion) with full names of keys; but using the key symbols solved the matter of width and make it cleaner.
So I made a new commit with @jagtalon's suggestions (thanks by  the way!): here's a new screenshot.

screenshot

**Comment by moollaza** [comment]
@jagtalon great suggestion! @Qeole nicely done with the update, this looks much better now and is certainly easier to read. I should have caught that!

**Comment by MrChrisW** [comment]
Looking good @Qeole 

@moollaza Should this IA provide a link to the [settings](https://duckduckgo.com/settings) page like the [help page](https://duck.co/help/features/keyboard-shortcuts) does? Enable or disable keyword settings.

**Comment by moollaza** [comment]
@MrChrisW I'm not sure how I feel about a link to the settings page. My only thought is if it will look out of place. But it obviously wouldn't hurt.

**Comment by MrChrisW** [comment]
@moollaza No problem, leave the link out. Is a cheat sheet template being worked on? I think I remember you mentioning it.

**Comment by zekiel** [comment]
@MrChrisW  it's in the pipeline but not the next thing. We're working on templates for goodies currently and a cheat sheet template should come after that.

**Comment by MrChrisW** [comment]
Thanks @zekiel, I saw the info about Goodie templates in the recent blog post. Looking forward to seeing that in use.

**Comment by jagtalon** [comment]
@Qeole nice nice! One thing I found is that I think it should take up as much space as it can on the right when on mobile:

screen shot 2015-02-06 at 4 26 46 am

**Comment by Qeole** [comment]
@jagtalon: That makes sense, thanks. The CSS rule for that was already in the stylesheet, but not at a place where it would have any effect; last commit moves it to the end of file, and this solves the issue.

screen shot 2015-02-06 at 17 47 32
For your information: while I was on it, I had a look at some other cheat sheets and noticed that the same issue arises with tmux cheat sheet goodie (not a surprise, since I took my CSS rules from this one), and that the gimp cheat sheet has overlapping columns on mobile displays (Crontab, regex and Vim all look fine).

**Comment by abeyang** [comment]
Hi @Qeole , thanks again for your contribution. I'm coming in a bit late to this party, but can we make a few more tweaks:

ddg-shortcuts-v1

Please note a few things:
- add a title :smile: 
- slightly bigger padding for the "bubbles", and darker (#e0e0e0)
- anything outside the bubbles (the "or"s) use a lighter color (#888888)
- moreover, let's try to keep the two columns consistent -- here's what the grid looks like:

http://cl.ly/image/1Z3m151X2h2z

10px between shortcut keys and description; 
56px between the two big columns.

Finally, can we try to have the mobile version line up as well? The shortcut keys columns should be consistent, as well as the description columns.

Whew, hopefully this wasn't too much! In any case, please take a stab at this, and @jagtalon and I can help you with any additional help (if you need it)!

**Comment by Qeole** [comment]
Thanks @abeyang for feedback!

So I had a look at your comments and here is what I got with last commit:

ddg4

Modifications include:
 * a title
 * padding for “bubble” keys
 * adapting layout to the grid you provided
 * I think the mobile version is alright now, please confirm whether this is what you expected

I'm not a CSS guru, but here are also some remarks:
 * adapting to your grid made me use absolute lengths (px) instead of relative, could you confirm this is what we want?
 * it also breaks the first two lines of shortcut keys in left column. The second (`Ctrl/⌘` + `↵` in particular looked better on a single row in my opinion (but *you* make decisions of course :-) ).
 * instant answer DIV got higher, I changed max-height from 45 to 52ex. Is it OK or should I switch back (and leave the scrollbar appear)? My opinion is that it's not so high and that it looks better without the scrollbar, but you tell me.
 * I did **NOT** change colors for bubbles and text (“or”s) because I did not set it at all (it uses style from `style.css` stylesheet for the SERP); if anything else, it would break style when users change colors in preferences, I'm quite sure that's not what we want here:

 ddg5

 Or if there is another simple way to do it (I don't know any darkening or lightening feature in CSS), please tell me!

Once again, any feedback is welcome.

**Comment by Qeole** [comment]
@abeyang: I'm realizing that I did not adapt correctly to your layout (I just understood that you expected the `Open results` and `Move around` headers to be inside the HTML table, which is not the case in my version).
I've made corresponding changes but haven't commited yet, because if the length of first columns are set by the length of their heading, I can't see how to align the description columns for mobile version. Would you have any hint? Or do you prefer the headings to remain outside the tables?

**Comment by abeyang** [comment]
Hi @Qeole , sorry but can't leave too long of a comment (going somewhere) -- I think it's ok if Open Results and Move Around are outside of the HTML table. Hope that helps...

**Comment by Qeole** [comment]
@abeyang Ok so I leave headings as(/where) they are now. Did you have some time to look at the modifications I mentioned in [response to your previous comment](956#issuecomment-74193901)?
Do the changes look correct, or is there anything more to modify?

**Comment by abeyang** [comment]
Thanks @Qeole . We'll take a look and get back to you! /cc @jagtalon

**Comment by Qeole** [comment]
Perfect, thanks :)

**Comment by moollaza** [comment]
@Qeole terribly sorry for the delay, this fell through the crack. Everything LGTM :+1: 

2015-03-19_16h59_27

**Comment by Qeole** [comment]
@moollaza Do not worry about the delay, there was no emergency here. Thanks to all of you for reviewing and merging!

**Comment by moollaza** [comment]
@Qeole Thanks to **you** for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by moollaza** [comment]
@Qeole congrats! It's live: https://duckduckgo.com/?q=DuckDuckGo+shortcuts&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by Qeole** [comment]
@moollaza I saw that a few days ago already, but thanks for the notice! I'd like to help again, but I don't have so much time for now. Maybe in some weeks, if I have interesting ideas to share :).

**Pull request merged**